subsys/net/lib/wifi_ready/wifi_ready.c

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <net/wifi_ready.h>
+#include <supp_events.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/net/net_event.h>
+
+LOG_MODULE_REGISTER(wifi_ready, CONFIG_WIFI_READY_EVENT_HANDLING_LOG_LEVEL);
+
+static wifi_ready_callback_t wifi_callbacks[CONFIG_MAX_CALLBACKS];
+static int callback_count;
+static K_MUTEX_DEFINE(wifi_mutex);
+
+#define WPA_SUPP_EVENTS (NET_EVENT_WPA_SUPP_READY) | \
+			(NET_EVENT_WPA_SUPP_NOT_READY)
+
+static struct net_mgmt_event_callback net_wpa_supp_cb;
+
+static void handle_wpa_supp_event(struct net_mgmt_event_callback *cb,
+				  struct net_if *iface, bool ready)
+{
+	if (ready) {
+		LOG_INF("Supplicant is ready for iface %p", iface);
+	} else {
+		LOG_INF("Supplicant is not ready for iface %p", iface);
+	}
+
+	k_mutex_lock(&wifi_mutex, K_FOREVER);
+	for (int i = 0; i < callback_count; i++) {
```

Comment:
how will this work in an array? This should be `ARRAY_SIZE(wifi_callbacks[])` (here and in all iterations).

E.g.,

add: cb1, cb2, cb3 -> count=3:
remove: cb2 -> count=2

then cb3 won't be invoked.

Reply:
you are doing `wifi_ready_callbacks[i] = wifi_ready_callbacks[callback_count - 1];`, but this still wont' work properly.

E.g., with only `cb1`, you won't delete the `cb1`, and keep calling it.